packages/valory/contracts/gnosis_safe/contract.py

```diff
@@ -774,3 +775,192 @@ def get_safe_txs(
+        ledger_api: EthereumApi,
+        contract_address: str,
+        removed_owner: Optional[str] = None,
+        from_block: BlockIdentifier = "earliest",
+        to_block: BlockIdentifier = "latest",
+    ) -> JSONLike:
+        """
+        Get all RemovedOwner events for a safe contract.
+
+        :param ledger_api: the ledger API object
+        :param contract_address: the contract address
+        :param removed_owner: the owner to check for, any owner qualifies if not provided.
+        :param from_block: from which block to search for events
+        :param to_block: to which block to search for events
+        :return: the added owner events
+        """
+        ledger_api = cast(EthereumApi, ledger_api)
+        safe_contract = cls.get_instance(ledger_api, contract_address)
+        entries = safe_contract.events.RemovedOwner.createFilter(
+            fromBlock=from_block,
+            toBlock=to_block,
+        ).get_all_entries()
+        if removed_owner is None:
+            removed_owner_events = list(
+                dict(
+                    tx_hash=entry.transactionHash.hex(),
+                    block_number=entry.blockNumber,
+                    owner=entry["args"]["owner"],
+                )
+                for entry in entries
+            )
+            return dict(
+                data=removed_owner_events,
+            )
+
+        checksummed_removed_owner = ledger_api.api.toChecksumAddress(removed_owner)
+        removed_owner_events = list(
+            dict(
+                tx_hash=entry.transactionHash.hex(),
+                block_number=entry.blockNumber,
+                owner=entry["args"]["owner"],
+            )
+            for entry in entries
+            if (
+                ledger_api.api.toChecksumAddress(entry["args"]["owner"])
+                == checksummed_removed_owner
+            )
+        )
+        return dict(
+            data=removed_owner_events,
+        )
+
+    @classmethod
+    def get_zero_transfer_events(
+        cls,
+        ledger_api: EthereumApi,
+        contract_address: str,
+        sender_address: str,
+        from_block: BlockIdentifier = "earliest",
+        to_block: BlockIdentifier = "latest",
+    ) -> JSONLike:
+        """
+        Get all zero transfer events from a given sender to the safe address.
+
+        :param ledger_api: the ledger API object
+        :param contract_address: the contract address
+        :param sender_address: the owner of the service, ie the address that triggers termination
+        :param from_block: from which block to search for events
+        :param to_block: to which block to search for events
+         :return: the zero transfer events
+        """
+        ledger_api = cast(EthereumApi, ledger_api)
+        safe_contract = cls.get_instance(ledger_api, contract_address)
+        sender_address = ledger_api.api.toChecksumAddress(sender_address)
+        entries = safe_contract.events.SafeReceived.createFilter(
```

Comment:
Isn't this event is responsible for tracking payments only to the safe multisig address?

Reply:
Correct. When the service owner makes a (0 value) transfer to the safe contract, we take that as the termination signal.

Reply:
and then we hand over control to the benevolent dictator?